### DISCUSSION TITLE
Problem with headers anchors when converting to gfm

### DISCUSSION THREAD

**Discussion by johnd0e** [jgm/pandoc#9794] [Q&A] [answered]
Consider such source sample:
```md
# Header 1 {#first}

[Link to header](#first)
```

As gfm does not support header attributes, this converts to:
```md
# Header 1

[Link to header](#first)
```

So the link is obviously broken.
I understand the gfm limitations, but it should be possible to fix this case automatically, replacing destination href with `#header-1`.

Is it possible to achieve it in some way now?

**Comment by tarleb** [comment]
It would certainly be possible to do with a (Lua) filter.

Another good solution would be to rely on pandoc's [implicit header references](https://pandoc.org/MANUAL#extension-implicit_header_references) feature. The above example would then be written as

``` markdown
# Header 1

[Link to header][Header 1]
```

Ensure that the right identifiers are used with `--from=markdown+gfm_auto_identifiers`.

**Comment by johnd0e** [comment]
> Another good solution would be to rely on pandoc's [implicit header references](https://pandoc.org/MANUAL#extension-implicit_header_references) feature.

Thanks, but I would like to keep my explicitly provided identifiers for other formats, and only use gfm auto identifiers for .md

> It would certainly be possible to do with a (Lua) filter.

Hmm, do you mean that I should implement this feature from scratch?
1. Enumerate headers, and keep list of original identifiers.
2. Enumerate links, and change their destinations to gfm-compatible identifiers.

Or there is some simpler way?

> Ensure that the right identifiers are used with --from=markdown+gfm_auto_identifiers.

Could you explain what is exact effect of "+gfm_auto_identifiers" if --from here?
Because I do not see any difference in result.

**Reply by tarleb** [reply]
> Hmm, do you mean that I should implement this feature from scratch?

Yes

> Or there is some simpler way?

None that I'm aware of

**Reply by johnd0e** [reply]
@tarleb 
I am trying to reimplement gfm headers identifiers in lua, but there is a problem: `string` lib functions are not unicode-aware, so it's impossible to do something like `local title = pandoc.utils.stringify(el):gsub("%p","")`.

Any suggestions?

**Reply by tarleb** [reply]
I believe the comments in #9591 should be helpful, esp. the linked StackOverflow Q&A.

**Reply by tarleb** [reply]
Reproducing the relevant Lua code here:

``` lua
local function make_id (inlines, via)
  local via = via or 'html'
  local heading = pandoc.Header(1, inlines)
  local temp_doc = pandoc.Pandoc{heading}
  local roundtripped_doc = pandoc.read(pandoc.write(temp_doc, via), via)
  return roundtripped_doc.blocks[1].identifier
end
```

The `via` parameter should be `html+gfm_auto_identifiers`  (or the like) to make sure that GFM identifiers are generated.

**Comment by johnd0e** [comment, marked as answer]
@tarleb 
Here is the resulting lua filter `mdHeadersLinks.lua`

```lua
-- adapted from https://stackoverflow.com/a/72600888/2520247
local temp_doc = pandoc.Pandoc{}
local function make_id (el, via)
  local via = via or "html"
  table.insert(temp_doc.blocks, pandoc.Header(1, el.content))
  local roundtripped_doc = pandoc.read(pandoc.write(temp_doc, via), via)
  local blocks = roundtripped_doc.blocks
  return blocks[#blocks].identifier
end

local headers = {}
local function Header (el)
  if el.identifier=="" then return end
  local id = make_id(el, "gfm")
  if id~=el.identifier then
    headers["#"..el.identifier] = "#"..id
  end
end

local function Link (el)
  local match = headers[el.target]
  if match then
    el.target = match
    return el
  end
end

return {
  { Header=Header },
  { Link=Link },
}
```
